### PR TITLE
Feature/signal registration

### PR DESCRIPTION
**Pull request by piiertho**
This implements signals and their registration.
Get back signals code from kotlin-native implem (maybe need to be centralized in a common artifact ?).
This fix mapping from JVM variant type to Wire type.
This adds `GodotError` class and fake `VariantArray`.
This does not implements signal binds, this needs real `VariantArray` implem.

### CODE REVIEW

**Review by chippmann — APPROVED** [review]
LGTM
I'm not so sure though if it makes sense to share the signal registration code. It's not much code and I think it might change a bit in the future and will diverge from the one we have in the kotlin native project

File: harness/tests/src/main/kotlin/godot/tests/Invocation.kt [hunk 1]
```diff
@@ -6,59 +6,112 @@ import godot.Spatial
...
+
+	override fun _onDestroy() {
+		invocation.free()
+	}
+
+	//Type cast checks
+	@RegisterFunction
+	fun parentIsSpatial() = getParent() is Spatial
+	@RegisterFunction
+	fun isObjectSpatial(obj: Object) = obj is Spatial
```
Comment (raniejade) [inline harness/tests/src/main/kotlin/godot/tests/Invocation.kt:116]: tabs :P

File: kt/godot-library/src/main/kotlin/godot/Wrapping.kt [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+/**
+ * Used for variadic icalls.
+ * TODO: Make non generic emit and connect methods at code generation time to avoid pattern matching when calling signals
+ */
+fun wrap(obj: Any?): KtVariant {
+    if (obj == null) {
+        return KtVariant(Unit)
+    }
+    return when (obj) {
```
Comment (raniejade) [inline kt/godot-library/src/main/kotlin/godot/Wrapping.kt:13]: I bet tristan is scratching his head when he sees this 😂

File: harness/tests/src/main/kotlin/godot/tests/Invocation.kt (hunk 1 above)
Comment (piiertho) [reply harness/tests/src/main/kotlin/godot/tests/Invocation.kt:116]: mmmh weird, still got spaces in intellij.

File: kt/godot-library/src/main/kotlin/godot/Wrapping.kt (hunk 2 above)
Comment (piiertho) [reply kt/godot-library/src/main/kotlin/godot/Wrapping.kt:13]: yep but we don't have much choice regarding variadic godot methods

Comment (piiertho) [reply kt/godot-library/src/main/kotlin/godot/Wrapping.kt:12]: Was not on all file.
Seems godot sets tabs each time I open project.

**Review by CedNaru — APPROVED** [review]
Okay for merge.
Note: Might be necessary to rework the string manipulation for the name properties in Signal.

**Comment by piiertho** [comment]
> Okay for merge.
> Note: Might be necessary to rework the string manipulation for the name properties in Signal.

I agree with this. There is maybe a way we can improve this like what we did with constructors. We should open an improvement issue.

**Pull request merged**